cabby/geo/walk.py

```diff
@@ -38,56 +52,304 @@ def compute_route(start_point: Point, end_point: Point, graph: nx.MultiDiGraph,
+
+    return None
+
+
+def get_pivot_near_goal(map: map_structure.Map, end_point: GeoDataFrame) -> Optional[Dict]:
+    '''Return a picked landmark near the end_point.
+    Arguments:
+      map: The map of a specific region.
+      end_point: The goal location.
+    Returns:
+      A single landmark near the goal location.
+    '''
+    try:
+        tags = {'name': True, 'wikidata': True,
+                'amenity': True, 'shop': True, 'tourism': True}
+        poi = ox.pois.pois_from_point(util.tuple_from_point(
+            end_point['centroid']), tags=tags, dist=40)
 
-def get_start_poi(poi, end_point):
-  pass
+        # Remove streets and roads.
+        poi = poi[poi['highway'].isnull()]
 
-def get_pivot(route):
-  pass
+    except Exception as e:
+        print(e)
+        return None
 
+    # Remove the endpoint.
+    nearby_poi = poi[poi['osmid'] != end_point['osmid']]
 
-def get_start_end_point():
+    prominent_poi = pick_prominent_pivot(nearby_poi)
 
-  # Read POI.
-  poi = load_poi()
+    if prominent_poi is None:
+        return
 
-  # Select end point.
-  end_point = get_end_poi(poi)
+    return prominent_poi.to_dict('records')[0]
 
-  # Select start point.
-  start_point = get_start_poi(poi, end_point)
 
-  # Compute route between start and end points.
-  route = compute_route()
+def get_pivots(route: GeoDataFrame, map: map_structure.Map, end_point: GeoDataFrame) -> Optional[Tuple[Dict, Dict]]:
+    '''Return a picked landmark on a given route.
+    Arguments:
+      route: Along the route a landmark will be chosen.
+      map: The map of a specific region.
+      end_point: The goal location.
+    Returns:
+      A single landmark.
+    '''
+
+    # Get POI along the route.
+    points_route = map.nodes[map.nodes['osmid'].isin(
+        route['osmid'])]['geometry']
+
+    try:
+        start_time = time.time()
+        poly = Polygon(points_route.tolist()).buffer(0.0005)
+        bounds = poly.bounds
+        bounding_box = box(bounds[0], bounds[1], bounds[2], bounds[3])
+        tags = {'wikipedia': True, 'wikidata': True, 'brand': True,
+                'tourism': True, 'amenity': True, 'shop': True, 'name': True}
+        df_pivots = ox.pois.pois_from_polygon(
+            bounding_box, tags=tags)
+
+    except Exception as e:
+        print(e)
+        return
+
+    # Polygon along the route.
+    df_pivots = df_pivots[df_pivots['geometry'].intersects(poly)]
+
+    # Remove streets.
+    df_pivots = df_pivots[(df_pivots['highway'].isnull())
+                          & (df_pivots['railway'].isnull())]
+
+    main_pivot = pick_prominent_pivot(df_pivots)
+    main_pivot = main_pivot.to_dict('records')[0]
+
+    # Get pivot near the goal location.
+    near_pivot = get_pivot_near_goal(map, end_point)
+    if main_pivot is None or near_pivot is None:
+        return
+
+    return main_pivot, near_pivot
+
+
+def get_points_and_route(map: map_structure.Map) -> Optional[Tuple[Dict, Dict, GeoDataFrame, Dict, Dict]]:
```

Comment:
When types get super complicated like this (and when we don't know what is inside things like Dicts), it's a sign that we need an object that holds these. Please create a data containing class that has clear named fields and types. (And probably good to use attrs style construction for it.)